File: java/src/test/java/neetcode/slidingwindow/p0003_longestsubstringwithoutrepeatingcharacters/SolutionTest.java
package neetcode.slidingwindow.p0003_longestsubstringwithoutrepeatingcharacters;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

class SolutionTest {

    private Solution sol;

    @BeforeEach
    void setUp() {
        sol = new Solution();
    }

    @Test
    void test_neetcode_example_one() {
        // "zxyzxyz" → "xyz" or "yzx" etc., length 3
        assertEquals(3, sol.lengthOfLongestSubstring("zxyzxyz"));
    }

    @Test
    void test_neetcode_example_two() {
        // "xxxx" → just "x", length 1
        assertEquals(1, sol.lengthOfLongestSubstring("xxxx"));
    }

    @Test
    void test_leetcode_classic() {
        // "abcabcbb" → "abc", length 3
        assertEquals(3, sol.lengthOfLongestSubstring("abcabcbb"));
    }

    @Test
    void test_all_unique() {
        // "abcdef" → whole string is the longest, length 6
        assertEquals(6, sol.lengthOfLongestSubstring("abcdef"));
    }

    @Test
    void test_empty_string() {
        assertEquals(0, sol.lengthOfLongestSubstring(""));
    }

    @Test
    void test_single_char() {
        assertEquals(1, sol.lengthOfLongestSubstring("a"));
    }

    @Test
    void test_pwwkew() {
        // Classic LeetCode example: longest is "wke", length 3
        // ("wkew" would be wrong because it has duplicate 'w')
        assertEquals(3, sol.lengthOfLongestSubstring("pwwkew"));
    }

    @Test
    void test_with_spaces_and_symbols() {
        // " " (single space) → length 1
        assertEquals(1, sol.lengthOfLongestSubstring(" "));
        // "ab c" → all 4 unique
        assertEquals(4, sol.lengthOfLongestSubstring("ab c"));
    }

    @Test
    void test_long_repeating_pattern() {
        // "dvdf" → "vdf", length 3
        // This catches a bug where left is advanced incorrectly past the duplicate
        assertEquals(3, sol.lengthOfLongestSubstring("dvdf"));
    }

    @Test
    void test_alternating_duplicates() {
        // "abba" → after the second 'b' the window shrinks; longest is "ab" or "ba", length 2
        // This catches a bug where left jumps too far back
        assertEquals(2, sol.lengthOfLongestSubstring("abba"));
    }
}

File: python/src/neetcode/slidingwindow/p0003_longestsubstringwithoutrepeatingcharacters/solution_test.py
import pytest
from solution import Solution, SolutionSet


@pytest.mark.parametrize("sol_class", [Solution, SolutionSet])
class TestBothSolutions:

    def test_classic_example(self, sol_class):
        assert sol_class().lengthOfLongestSubstring("abcabcbb") == 3

    def test_all_same(self, sol_class):
        assert sol_class().lengthOfLongestSubstring("bbbbb") == 1

    def test_neetcode_example(self, sol_class):
        assert sol_class().lengthOfLongestSubstring("zxyzxyz") == 3

    def test_empty_string(self, sol_class):
        assert sol_class().lengthOfLongestSubstring("") == 0

    def test_single_char(self, sol_class):
        assert sol_class().lengthOfLongestSubstring("a") == 1

    def test_all_unique(self, sol_class):
        assert sol_class().lengthOfLongestSubstring("abcdef") == 6

    def test_pwwkew(self, sol_class):
        assert sol_class().lengthOfLongestSubstring("pwwkew") == 3

    def test_dvdf(self, sol_class):
        # Tests correct left pointer handling
        assert sol_class().lengthOfLongestSubstring("dvdf") == 3

    def test_abba(self, sol_class):
        assert sol_class().lengthOfLongestSubstring("abba") == 2
